### PR TITLE
Page loading indicator

### LINKED ISSUE #354: Page loading indicators
- Transaction sending should show a spinning progress indicator while attempting the send. Currently nothing is shown. See android.
- history page, No loading indicator is shown while transactions are being loaded. see android.
- settings, wallet log takes a very long time to load and shows no indicator to the user.

### PR DESCRIPTION
**Pull request by macsleven**
closes #354

### CODE REVIEW

File: decred_wallet/Features/History/TransactionHistoryViewController.swift [hunk 1]
```diff
@@ -75,23 +77,32 @@ class TransactionHistoryViewController: UIViewController {
...
         self.Filtercontent.removeAll()
-        DispatchQueue.global(qos: .background).async { [weak self] in
-            guard let this = self else { return }
+        DispatchQueue.global(qos: .background).async {
             do {
                 var getTxsError: NSError?
                 // use limit = 0 to return all transactions
                 let jsonResponse = AppDelegate.walletLoader.wallet?.getTransactions(0, error: &getTxsError)
                 if getTxsError != nil {
+                    DispatchQueue.main.async {
```
Comment (itswisdomagain) [inline decred_wallet/Features/History/TransactionHistoryViewController.swift]: This isn't needed since the catch handler below already does this.

File: decred_wallet/Features/History/TransactionHistoryViewController.swift [hunk 2]
```diff
@@ -238,3 +249,4 @@ extension TransactionHistoryViewController: UITableViewDataSource, UITableViewDe
         return cell
     }
 }
+
```
Comment (itswisdomagain) [inline decred_wallet/Features/History/TransactionHistoryViewController.swift]: Why do we need this extra line?

File: decred_wallet/Features/History/TransactionHistoryViewController.swift [hunk 3]
```diff
@@ -57,6 +58,7 @@ class TransactionHistoryViewController: UIViewController {
         }
         self.visible = true
         if (self.FromMenu){
+            refreshControl.programaticallyBeginRefreshing(in: self.tableView)
```
Comment (itswisdomagain) [inline decred_wallet/Features/History/TransactionHistoryViewController.swift]: Do we only need this loading indicator when this page is accessed from menu?

File: decred_wallet/Extensions/Extensions.swift [hunk 4]
```diff
@@ -69,3 +69,11 @@ extension Date {
         self = Date(timeIntervalSince1970: TimeInterval(milliseconds) / 1000)
     }
 }
+
+extension UIRefreshControl {
+    func programaticallyBeginRefreshing(in tableView: UITableView) {
```
Comment (itswisdomagain) [inline decred_wallet/Extensions/Extensions.swift]: `programaticallyBeginRefreshing`? Is there any executed swift code that is not "programatically" executed? I'm quite sure you can use `beginRefreshing` as an overloaded method.

File: decred_wallet/Extensions/Extensions.swift [hunk 5]
```diff
@@ -69,3 +69,11 @@ extension Date {
...
+        beginRefreshing()
```
Comment (itswisdomagain) [inline decred_wallet/Extensions/Extensions.swift]: `self.beginRefreshing()` please, so one does not begin to wonder where this `beginRefreshing` function is written.

File: decred_wallet/Extensions/Extensions.swift [hunk 6]
```diff
@@ -69,3 +69,11 @@ extension Date {
...
+        let offsetPoint = CGPoint.init(x: 0, y: -frame.size.height)
```
Comment (itswisdomagain) [inline decred_wallet/Extensions/Extensions.swift]: Please add a comment as to why this is necessary.

File: decred_wallet/Features/History/TransactionHistoryViewController.swift [hunk 7]
```diff
@@ -37,6 +37,7 @@ class TransactionHistoryViewController: UIViewController {
         initFilterBtn()
         self.tableView.addSubview(self.refreshControl)
         self.tableView.register(UINib(nibName: TransactionTableViewCell.identifier, bundle: nil), forCellReuseIdentifier: TransactionTableViewCell.identifier)
+        self.tableView.tableFooterView = UIView.init()
```
Comment (itswisdomagain) [inline decred_wallet/Features/History/TransactionHistoryViewController.swift]: Why is this necessary?

File: decred_wallet/Features/History/TransactionHistoryViewController.swift [hunk 8]
```diff
@@ -75,23 +77,32 @@ class TransactionHistoryViewController: UIViewController {
     
     @objc func handleRefresh(_ refreshControl: UIRefreshControl) {
         self.prepareRecent()
-        refreshControl.endRefreshing()
+
```
Comment (itswisdomagain) [inline decred_wallet/Features/History/TransactionHistoryViewController.swift]: Unnecessary empty line space.

File: decred_wallet/view_controller/SendViewController.swift [hunk 9]
```diff
@@ -346,19 +346,22 @@ class SendViewController: UIViewController, UITextFieldDelegate,UITextPasteDeleg
...
                     
                     let isShouldBeConfirmed = UserDefaults.standard.bool(forKey: "pref_spend_fund_switch")
                     let result = try self.wallet?.sendTransaction(password.data(using: .utf8), destAddr: walletAddress, amount: Int64(amount) , srcAccount: account , requiredConfs: isShouldBeConfirmed ? 0 : 2, sendAll: sendAll ?? false)
                     
                     DispatchQueue.main.async {
+                        progressHud.dismiss()
                         self.transactionSucceeded(hash: result?.hexEncodedString())
                         return
                     }
                     return
```
Comment (itswisdomagain) [inline decred_wallet/view_controller/SendViewController.swift]: This return is useless, so is the next line that is empty.

File: decred_wallet/view_controller/WalletLogViewController.swift [hunk 10]
```diff
@@ -7,13 +7,17 @@
 // license that can be found in the LICENSE file.
 
 import UIKit
+import JGProgressHUD
 
 class WalletLogViewController: UIViewController {
     @IBOutlet weak var logTextView: UITextView!
-    
+    static var progressHud = JGProgressHUD(style: .light)
```
Comment (itswisdomagain) [inline decred_wallet/view_controller/WalletLogViewController.swift]: - Don't make this static please. Dismiss the hud in `viewDidLoad`, after reading the log.
- Declare but don't assign here. Assign in `viewDidLoad`

File: decred_wallet/view_controller/WalletLogViewController.swift [hunk 11]
```diff
@@ -7,13 +7,17 @@
...
     override func viewDidLoad() {
         super.viewDidLoad()
         self.navigationItem.title = "Wallet Log"
+        WalletLogViewController.progressHud = Utils.showProgressHud(withText: "Loading log...")
+    }
+    override func viewDidAppear(_ animated: Bool) {
         self.logTextView.text = WalletLogViewController.readLog()
```
Comment (itswisdomagain) [inline decred_wallet/view_controller/WalletLogViewController.swift]: Why are we reading the log in `viewDidAppear` instead of in `viewDidLoad`?

Comment (macsleven) [reply decred_wallet/view_controller/WalletLogViewController.swift]: For better transition when moving into the controller.

File: decred_wallet/Features/History/TransactionHistoryViewController.swift (hunk 7 above)
Comment (macsleven) [reply decred_wallet/Features/History/TransactionHistoryViewController.swift]: This removes empty lines on tableview when there is no data on the list

File: decred_wallet/Features/History/TransactionHistoryViewController.swift (hunk 3 above)
Comment (macsleven) [reply decred_wallet/Features/History/TransactionHistoryViewController.swift]: Yes, swipe down refresh handles the loading indicator in the app

File: decred_wallet/Extensions/Extensions.swift (hunk 6 above)
Comment (macsleven) [reply decred_wallet/Extensions/Extensions.swift]: This is for displaying the loading indicator without the user swiping down the tableview

File: decred_wallet/Extensions/Extensions.swift (hunk 4 above)
Comment (macsleven) [reply decred_wallet/Extensions/Extensions.swift]: Renamed to showLoader

File: decred_wallet/view_controller/WalletLogViewController.swift [hunk 12]
```diff
@@ -7,29 +7,36 @@
...
 
 class WalletLogViewController: UIViewController {
     @IBOutlet weak var logTextView: UITextView!
-    
+    var progressHud = JGProgressHUD(style: .light)
     override func viewDidLoad() {
         super.viewDidLoad()
         self.navigationItem.title = "Wallet Log"
-        self.logTextView.text = WalletLogViewController.readLog()
+       self.progressHud = Utils.showProgressHud(withText: "Loading log...")
```
Comment (itswisdomagain) [inline decred_wallet/view_controller/WalletLogViewController.swift]: Correct indentation.

File: decred_wallet/Features/History/TransactionHistoryViewController.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (macsleven) [reply decred_wallet/Features/History/TransactionHistoryViewController.swift]: its not a hack i added a footer view to it which is normal for a table view, which prevent the tableview from showing empty rows

File: decred_wallet/Extensions/Extensions.swift [hunk 13]
```diff
@@ -69,3 +69,12 @@ extension Date {
         self = Date(timeIntervalSince1970: TimeInterval(milliseconds) / 1000)
     }
 }
+
+extension UIRefreshControl {
+    //display loading indicator without the aid of user swipping down tableview
```
[the remark this answers is not in the document]
Comment (macsleven) [reply decred_wallet/Extensions/Extensions.swift]: Did you read the comment below about IOS 11?

File: decred_wallet/view_controller/WalletLogViewController.swift (hunk 12 above)
Comment (macsleven) [reply decred_wallet/view_controller/WalletLogViewController.swift]: done

File: decred_wallet/Extensions/Extensions.swift (hunk 13 above)
Comment (itswisdomagain) [inline decred_wallet/Extensions/Extensions.swift]: Some other dev seeing this code might be tempted to fix it, it doesn't seem like the most logical or straightforward way to display a loading indicator. Your comment should include why you're using this approach and probably other approaches you considered and why they would not work. May also include the link to SO comment/answer that guided this your implementation.

See https://medium.freecodecamp.org/code-comments-the-good-the-bad-and-the-ugly-be9cc65fbf83

File: decred_wallet/view_controller/WalletLogViewController.swift [hunk 14]
```diff
@@ -7,29 +7,36 @@
...
     @IBOutlet weak var logTextView: UITextView!
-    
+    var progressHud = JGProgressHUD(style: .light)
     override func viewDidLoad() {
         super.viewDidLoad()
         self.navigationItem.title = "Wallet Log"
-        self.logTextView.text = WalletLogViewController.readLog()
+        self.progressHud = Utils.showProgressHud(withText: "Loading log...")
+    }
+    override func viewDidAppear(_ animated: Bool) {
```
Comment (itswisdomagain) [inline decred_wallet/view_controller/WalletLogViewController.swift]: Separate these 2 functions with a single line space.

File: decred_wallet/Features/History/TransactionHistoryViewController.swift (hunk 7 above)
Comment (itswisdomagain) [inline decred_wallet/Features/History/TransactionHistoryViewController.swift]: > its not a hack i added a footer view to it which is normal for a table view, which prevent the tableview from showing empty rows

Table view footers are not meant to "prevent the tableview from showing empty rows". Can you do this instead? It's a more straightforward solution.
> You should check if there's no data to display and remove the table view separator.

File: decred_wallet/Extensions/Extensions.swift (hunk 13 above)
Comment (macsleven) [reply decred_wallet/Extensions/Extensions.swift]: UIRefreshControl has its loading indicator which is used for tableview swipe down refresh, reusing  same loader in this case seems straightforward.

File: decred_wallet/Features/History/TransactionHistoryViewController.swift (hunk 7 above)
Comment (macsleven) [reply decred_wallet/Features/History/TransactionHistoryViewController.swift]: i will take it off, seems there is a different issue created for this #317
